astacus/node/cassandra.py

```diff
@@ -0,0 +1,156 @@
+"""
+
+Copyright (c) 2021 Aiven Ltd
+See LICENSE for details
+
+Cassandra handling that is run on every node in the Cluster
+
+"""
+
+from .node import NodeOp
+from astacus.common import ipc
+from astacus.common.cassandra.client import CassandraClient
+from astacus.common.cassandra.config import SNAPSHOT_NAME
+from astacus.common.cassandra.schema import CassandraSchema
+from astacus.common.cassandra.utils import is_system_keyspace
+from astacus.common.exceptions import TransientException
+from typing import List, Optional
+
+import contextlib
+import logging
+import shutil
+import subprocess
+import tempfile
+import yaml
+
+logger = logging.getLogger(__name__)
+
+SNAPSHOT_GLOB = f"data/*/*/snapshots/{SNAPSHOT_NAME}"
+
+
+# Clients do not really provide the sub-requests, but it is cleaner to
+# have simply single class handling all Cassandra sub-requests
+class CassandraRequest(ipc.NodeRequest):
+    subop: ipc.CassandraSubOp
+    tokens: Optional[List[str]]
+
+
+class CassandraOp(NodeOp):
+    def start(self, *, req):
+        subop = req.subop
+        if subop == ipc.CassandraSubOp.get_schema_hash:
+            self.result = ipc.CassandraGetSchemaHashResult(schema_hash="")
+        self.req = req
+        logger.debug("start_cassandra %r", req)
+        assert self.config.cassandra
+        return self.start_op(
+            op_name="cassandra",
+            op=self,
+            fun={
+                ipc.CassandraSubOp.get_schema_hash: self.get_schema_hash,
+                ipc.CassandraSubOp.remove_snapshot: self.remove_snapshot,
+                ipc.CassandraSubOp.restore_snapshot: self.restore_snapshot,
+                ipc.CassandraSubOp.start_cassandra: self.start_cassandra,
+                ipc.CassandraSubOp.stop_cassandra: self.stop_cassandra,
+                ipc.CassandraSubOp.take_snapshot: self.take_snapshot,
+            }[subop],
+        )
+
+    def _get_schema_hash(self):
+        with CassandraClient(self.config.cassandra.client).connect() as cas:
+            schema = CassandraSchema.from_cassandra_session(cas)
+            return schema.calculate_hash()
```

Comment:
Yes, but the hash calculation is (almost) one-liner and at least then it matches exactly what we want to restore/backup; I'm not sure which is preferable. Also, Cashew did it this way; we can easily enough change it though if you think the official schema is better though.

Reply:
since this is used to validate that schema has not changed while we're preparing the manifest, I think using the "official" version would be more appropriate. The assumption being that if the "official" version has not changed, then there were no changes we really need to care about. It also might help if half the nodes are on one version of Astacus and half on another, and some order-influencing change occured that prevent the hashes to match, while the contents of the schema remain the same.

Reply:
That's fair - changed to use the official hash.